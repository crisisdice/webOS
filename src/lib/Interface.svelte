<style>
  input {
    left: -1000px;
    position: absolute;
  }

  .caret-empty {
    animation: empty 1s step-end infinite;
    color: transparent;
  }

  .caret-full {
    animation: full 1s step-end infinite;
    color: #fff;
  }

  @keyframes empty {
    50% {
      background-color: #fff;
    }
  }

  @keyframes full {
    50% {
      background-color: #fff;
      color: #000;
    }
  }
</style>

<script lang="ts">
  import Prompt from './Prompt.svelte'

  const EMPTY = ''
  const DASH = '_'

  let oldCmds: { cmd: string; wd: string; stdout: string }[] = []
  let precaret = ''
  let caret = DASH
  let postcaret = ''
  let caretClass = 'caret-empty'
  let CONTROL_DOWN = false
  let prevCmd = 0

  let pwd = '/home/guest'

  let structure = {
    'home': {
      'guest': {
        '.bashrc': '#!/usr/sh',
        '.history': ''
      },
      'root': {}
    }
  }

  let user = 'guest'
  let pwdObj = structure['home'][user]

  const echo = (args: string) => {
    return args.replaceAll('"', '')
  }

  const ls = () => {
    return Object.keys(pwdObj).join(' ')
  }

  const cat = (arg: string) => {
    const stat = () => {
      if (!Object.keys(pwdObj).find(k => k === arg)) return `cat : ${arg}: No such file or directory`
      return pwdObj[arg]
    }
    const file = stat()
    return (typeof file === 'string') 
      ? file
      : `cat: ${arg}: Is a directory`
  }

  const mkdir = (arg: string) => {
    if (!!pwdObj[arg]) return `mkdir: ${arg}: File exists`
    pwdObj[arg] = {}
    return
  }

  const cd = (args: string) => {
    if (args === EMPTY || args === '~') {
      pwdObj = structure['home'][user]
      pwd = `/home/${user}`
      return
    }

    if (args === '/') {
      pwd = '/'
      pwdObj = structure
      return
    }

    let steps = args[0] === '/' ? args.split('/') : `${pwd === '/' ? '' : pwd }/${args}`.split('/')
    let tmp = structure

    if (args === '..') {
      // TODO all relative
      const pwdTmp = pwd.split('/')
      steps = pwdTmp.slice(0, pwdTmp.length - 1)
    }

    for (const step of steps) {
      if (step === "") continue
      tmp = tmp[step]
      if (!tmp) return `cd: no such file or directory: ${args}`
    }

    pwdObj = tmp
    pwd = (steps.length === 1 && steps[0] === "") || steps.length === 0 ? '/' : steps.join('/')
  }

  const evaluate = (input: string) => {
    if (input === EMPTY) return

    const cmds = {
      echo,
      cd,
      ls,
      pwd: 'dummy',
      cat,
      mkdir
    }

    // TODO command parsing
    const cmd = (input.split(' ')?.[0] ?? input).trim()

    const argsI = input.indexOf(' ')
    const args = argsI === -1 ? EMPTY : input.slice(argsI + 1)

    if (!cmds[cmd]) return `sh: command not found: ${cmd}`

    if (cmd === 'pwd') return `${pwd}/`

    return cmds[cmd](args)
  }

  const up = ({ key }: KeyboardEvent) => {
    console.log({ key })

    /* check for ctl codes */
    if (CONTROL_DOWN) {
      switch (key) {
        case "l":
          oldCmds = []
          precaret = postcaret = EMPTY
          caret = DASH
          return
        case "Control":
          CONTROL_DOWN = false
          return
        default:
          return
      }
    }

    switch (key) {
      case "Enter": {
        const cmd = caret === DASH ? `${precaret}${postcaret}` : `${precaret}${caret}${postcaret}`
        const wd = pwd
        oldCmds = [...oldCmds, { cmd, stdout: evaluate(cmd), wd }]
        precaret = postcaret = EMPTY
        caret = DASH
        prevCmd = 0
        return
      }
      case "ArrowLeft": {
        if (precaret === EMPTY) return
        const lastIndex = precaret.length - 1
        const last = precaret[lastIndex]
        precaret = precaret.slice(0, lastIndex)
        postcaret = caret === DASH ? EMPTY : `${caret}${postcaret}`
        caret = last
        caretClass = 'caret-full'
        return
      }
      case "ArrowUp": {
        postcaret = EMPTY
        caret = DASH
        const history = oldCmds.filter((c) => c.cmd !== EMPTY).reverse()
        if (prevCmd < history.length) {
          precaret = history[prevCmd].cmd
          prevCmd += 1
        }
        return
      }
      case "ArrowDown": {
        postcaret = EMPTY
        caret = DASH
        const history = oldCmds.filter((c) => c.cmd !== EMPTY).reverse()
        if (prevCmd > -1) {
          prevCmd -= 1
        }
        precaret = prevCmd === -1 ? EMPTY : history[prevCmd].cmd
        if (prevCmd === -1) prevCmd = 0
        return
      }
      case "ArrowRight": {
        if (postcaret === EMPTY) {
          if (caret !== DASH) {
            precaret = `${precaret}${caret}`
          }
          caret = DASH
          caretClass = 'caret-empty'
          return
        }
        const first = postcaret[0]
        precaret = `${precaret}${caret}`
        caret = first
        postcaret = postcaret.slice(1, postcaret.length)
        return
      }
      case "Backspace": {
        precaret = precaret.slice(0, precaret.length - 1)
        return
      }
      case "Shift":
      case "Tab":
        return
      default: {
        precaret += key
      }
    }
  }

  const down = ({ key }: KeyboardEvent) => {
    switch (key) {
      case "Control":
        CONTROL_DOWN = true
        return
      default:
        return
    }
  }

  const refocus = (e: Event) => window.setTimeout(() => (e.target as HTMLInputElement).focus(), 0)
</script>

<div>
  {#each oldCmds as oldCmd}
    <Prompt pwd={oldCmd.wd} usr={user}>{oldCmd.cmd}</Prompt>
    {#if oldCmd.stdout}<span>{oldCmd.stdout}</span>{/if}
  {/each}
  <Prompt pwd={pwd} usr={user}>
    {precaret}<span class={caretClass}>{caret}</span>{postcaret}
  </Prompt>
  <!-- svelte-ignore a11y-autofocus -->
  <input
    autofocus
    on:keydown={down}
    on:keyup={up}
    on:blur={refocus}
  />
</div>
