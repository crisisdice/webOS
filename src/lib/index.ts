/* kernel */
export * from './fs'

/* binaries */
export * from './vi'
export * from './sh'
