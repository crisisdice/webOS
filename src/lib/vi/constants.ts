export enum VI_MODE {
  VISUAL,
  INSERT,
  COMMAND,
  VISUAL_BLOCK,
  VISUAL_LINE,
}
