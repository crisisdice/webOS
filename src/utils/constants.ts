export const EMPTY = ''

export const ENV = 'ENV'

export const FS = 'FS'

export const PWD = 'PWD'

export const HOME = 'HOME'

export const ARROW_KEYS = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight']
