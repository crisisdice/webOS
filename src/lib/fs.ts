import { FS, ENV } from './constants'
import type { ENV as IENV, Directory, Stat } from './types'

function get<T>(lsKey: 'ENV' | 'FS'): T | null {
  const ls = localStorage.getItem(lsKey)
  if (!ls) return null
  return JSON.parse(ls) as T
}

const absoluteTokens = (path: string) => path.split('/').filter((t) => !!t)

function parsePath(path: string): string[] {
  if (path === '') throw new Error('no path')
  if (path[0] === '/') return absoluteTokens(path)

  const pwdTokens = absoluteTokens(getEnv('PWD'))

  const pathTokens = path.split('/').filter((t) => !!t)

  let stack = [...pwdTokens]

  for (const token of pathTokens) {
    if (token === '.') {
      continue
    }

    if (token === '~') {
      stack = absoluteTokens(getEnv('HOME'))
      continue
    }

    if (token === '..') {
      if (stack.pop() === undefined) {
        throw new Error('path error')
      }
      continue
    }

    stack.push(token)
  }

  return stack
}

function traverse(tokens: string[], fs: Directory) {
  let tmp = fs

  for (const token of tokens) {
    tmp = tmp[token] as Directory
    if (tmp === undefined) throw new Error('dir does not exist')
  }

  return tmp
}

export function setFs(fs: Directory): void {
  localStorage.setItem(FS, JSON.stringify(fs))
}

export function getEnv(key: keyof IENV): string {
  return get<IENV>(ENV)[key] ?? ''
}

export function setEnv(key: string, value: string): void {
  const env = get<IENV>(ENV) ?? {}
  localStorage.setItem(ENV, JSON.stringify({ ...env, [key]: value }))
}

export function write({
  path,
  name,
  obj,
}: {
  path: string
  name: string
  obj: string | Directory | null
}) {
  const fs = get<Directory>(FS)
  const location = traverse(parsePath(path), fs)

  if (obj === null) {
    delete location[name]
  } else {
    location[name] = obj
  }

  setFs(fs)
}

export function stat(args: string): Stat {
  try {
    const fs = get<Directory>(FS)
    const tokens = parsePath(args === '' ? getEnv('PWD') : args)
    const path = '/' + tokens.join('/')
    const obj = traverse(tokens, fs)
    return { exists: true, path, isDirectory: typeof obj !== 'string', obj }
  } catch (e) {
    return { exists: false, path: '-', isDirectory: false, obj: '-' }
  }
}
