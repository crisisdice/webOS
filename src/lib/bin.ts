import { getEnv, setEnv, stat, write } from './fs'

const EMPTY = ''

const echo = (args: string) => {
  return args.replaceAll('"', '')
}

const ls = (args: string) => {
  const { exists, isDirectory, obj } = stat(args)
  if (!exists) return `ls: cannot access '${args}': No such file or directory`
  return isDirectory ? Object.keys(obj).join(' ') : args.split('/').at(-1)
}

const cat = (args: string): string => {
  const { exists, isDirectory, obj } = stat(args)
  if (!exists) return `cat: cannot access '${args}': No such file or directory`
  return isDirectory ? `cat: ${args}: Is a directory` : (obj as string)
}

const cd = (args: string) => {
  const { exists, path } = stat(args)
  if (!exists) return `cd: no such file or directory: ${args}`
  setEnv('PWD', path)
}

const rm = (args: string) => {
  const { exists, path } = stat(args)
  if (!exists) return `rm: no such file or directory: ${args}`
  write({ ...separate(path), obj: null })
}

const touch = (args: string) => {
  const { exists } = stat(args)
  if (exists) return ''
  write({ ...separate(args), obj: '' })
}

const separate = (args: string): { name: string; path: string } => {
  if (args.indexOf('/') === -1) return { path: '.', name: args }
  const tokens = args.split('/')
  const path = tokens.slice(0, tokens.length - 1).join('/')
  return { name: tokens.at(-1), path }
}

const node = (args: string) => {
  const { exists, obj, isDirectory } = stat(args)
  if (isDirectory) return `node: ${args}: is a directory`

  try {
    let stdout = ''

    const tmp = window.console

    // TODO parse \n as a new span
    window.console = {
      log: (args: string) => {
        stdout += `${args}\n`
      },
    } as Console

    eval(exists ? (obj as string) : args)

    window.console = tmp
    console.log('done')

    return stdout
  } catch (e) {
    return (e as Error).message
  }
}

export const evaluate = (input: string) => {
  if (input === EMPTY) return

  const cmds: Record<string, (args: string) => string> = {
    echo,
    cd,
    ls,
    pwd: () => 'dummy',
    cat,
    rm,
    touch,
    node,
  }

  // TODO command parsing, .i.e. flags, different args, shell expansions, pipes, quotes
  const cmd = (input.split(' ')?.[0] ?? input).trim()
  const argsI = input.indexOf(' ')
  const args = argsI === -1 ? EMPTY : input.slice(argsI + 1)

  if (!cmds[cmd]) return `sh: command not found: ${cmd}`
  if (cmd === 'pwd') return `${getEnv('PWD')}/`

  if (args.indexOf('>') > -1) {
    const tokens = args.split('>')
    const stdout = cmds[cmd](tokens[0].trim())
    write({ ...separate(tokens[1].trim()), obj: stdout })
    return
  }

  return cmds[cmd](args)
}
