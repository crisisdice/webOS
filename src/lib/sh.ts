import { getEnv, setEnv, setFs, write } from './fs'
import { separate } from './utils'
import { EMPTY, PWD, HOME } from './constants'
import { echo, cd, ls, cat, rm, touch } from './bin'

export const init = () => {
  setEnv(PWD, '/home/guest')
  setEnv(HOME, '/home/guest')

  setFs({
    home: {
      guest: {
        '.bashrc': '#!/usr/sh',
        '.history': '',
      },
      root: {},
    },
  })
  console.log('initalized')
}

export const evaluate = (input: string) => {
  if (input === EMPTY) return

  const cmds: Record<string, (args: string) => string> = {
    echo,
    cd,
    ls,
    pwd: () => 'dummy',
    cat,
    rm,
    touch,
  }

  // TODO command parsing, .i.e. flags, different args, shell expansions, pipes, quotes
  // TODO and then pass args as string[]
  const cmd = (input.split(' ')?.[0] ?? input).trim()
  const argsI = input.indexOf(' ')
  const args = argsI === -1 ? EMPTY : input.slice(argsI + 1)

  if (!cmds[cmd]) return `sh: command not found: ${cmd}`
  if (cmd === 'pwd') return `${getEnv('PWD')}/`

  if (args.indexOf('>') > -1) {
    const tokens = args.split('>')
    const stdout = cmds[cmd](tokens[0].trim())
    write({ ...separate(tokens[1].trim()), obj: stdout })
    return
  }

  return cmds[cmd](args)
}
