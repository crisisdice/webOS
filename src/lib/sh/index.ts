import { getEnv, write } from '../fs'
import { separate } from '../path'
import { echo, cd, ls, cat, rm, touch } from '../bin'
import { EMPTY } from '../../utils'

export const evaluate = (input: string) => {
  if (input === EMPTY) return

  const cmds: Record<string, (args: string) => string> = {
    echo,
    cd,
    ls,
    pwd: () => 'dummy',
    cat,
    rm,
    touch,
  }

  // TODO command parsing, .i.e. flags, different args, shell expansions, pipes, quotes
  // TODO and then pass args as string[]
  const cmd = (input.split(' ')?.[0] ?? input).trim()
  const argsI = input.indexOf(' ')
  const args = argsI === -1 ? EMPTY : input.slice(argsI + 1)

  if (!cmds[cmd]) return `sh: command not found: ${cmd}`
  if (cmd === 'pwd') return `${getEnv('PWD')}/`

  if (args.indexOf('>') > -1) {
    const tokens = args.split('>')
    const stdout = cmds[cmd](tokens[0].trim())
    write({ ...separate(tokens[1].trim()), obj: stdout })
    return
  }

  return cmds[cmd](args)
}
