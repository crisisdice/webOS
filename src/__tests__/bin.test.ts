import { run } from './helpers'
import { ENTER } from './constants'
import { init } from '../lib/sh'

const tests = [
  [...'echo hello'.split(''), ENTER],
  [...'cd /'.split(''), ENTER],
  [...'cd ~'.split(''), ENTER],
  [...'cd ../root'.split(''), ENTER],
  [...'touch hello.txt'.split(''), ENTER],
  [...'cat "hello" > hello.txt'.split(''), ENTER],
  [...'echo "hello" > hello.txt'.split(''), ENTER],
  [...'echo "hello" > ../hello.txt'.split(''), ENTER],
  [...'rm .bashrc'.split(''), ENTER],
  [...'rm ../root'.split(''), ENTER],
  [...'ls'.split(''), ENTER],
  [...'ls /'.split(''), ENTER],
]

beforeEach(() => init())

for (const sequence of tests) {
  run(sequence)
}
