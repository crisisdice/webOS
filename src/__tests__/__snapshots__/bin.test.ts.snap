// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cat "hello" > hello.txt<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "cat "hello" > hello.txt",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`cat "hello" > hello.txt<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":"","hello.txt":"cat: cannot access '\\"hello\\"': No such file or directory"},"root":{}}}"`;

exports[`cat "hello" > hello.txt<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`cd ../root<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "cd ../root",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/root",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`cd ../root<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`cd ../root<ENTER> 3`] = `"{"PWD":"/home/root","HOME":"/home/guest"}"`;

exports[`cd /<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "cd /",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`cd /<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`cd /<ENTER> 3`] = `"{"PWD":"/","HOME":"/home/guest"}"`;

exports[`cd ~<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "cd ~",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`cd ~<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`cd ~<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`echo "hello" > ../hello.txt<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "echo "hello" > ../hello.txt",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`echo "hello" > ../hello.txt<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{},"hello.txt":"hello"}}"`;

exports[`echo "hello" > ../hello.txt<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`echo "hello" > hello.txt<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "echo "hello" > hello.txt",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`echo "hello" > hello.txt<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":"","hello.txt":"hello"},"root":{}}}"`;

exports[`echo "hello" > hello.txt<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`echo hello<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "echo hello",
      "stdout": "hello",
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`echo hello<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`echo hello<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`ls /<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "ls /",
      "stdout": "home",
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`ls /<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`ls /<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`ls<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "ls",
      "stdout": ".bashrc .history",
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`ls<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`ls<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`rm ../root<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "rm ../root",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`rm ../root<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""}}}"`;

exports[`rm ../root<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`rm .bashrc<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "rm .bashrc",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`rm .bashrc<ENTER> 2`] = `"{"home":{"guest":{".history":""},"root":{}}}"`;

exports[`rm .bashrc<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`touch hello.txt<ENTER> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "FULL_SCREEN": false,
  "NAME": "sh",
  "OLD_COMMANDS": [
    {
      "cmd": "touch hello.txt",
      "stdout": undefined,
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`touch hello.txt<ENTER> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":"","hello.txt":""},"root":{}}}"`;

exports[`touch hello.txt<ENTER> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;
